pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -387,11 +376,19 @@ func (og *operationGenerator) GenerateAttachVolumeFunc(
 		return nil, nil
 	}
 
+	eventRecorderFunc := func(err *error) {
+		if *err != nil {
+			for _, pod := range volumeToAttach.ScheduledPods {
+				og.recorder.Eventf(pod, v1.EventTypeWarning, kevents.FailedAttachVolume, (*err).Error())
+			}
+		}
+	}
+
 	return volumetypes.GeneratedOperations{
 		OperationFunc:     attachVolumeFunc,
 		EventRecorderFunc: eventRecorderFunc,
-		CompleteFunc:      util.OperationCompleteHook(util.GetFullQualifiedPluginNameForVolume(attachableVolumePlugin.GetPluginName(), volumeToAttach.VolumeSpec), "volume_attach"),
-	}, nil
+		CompleteFunc:      util.OperationCompleteHook(util.GetFullQualifiedPluginNameForVolume(attachableVolumePluginName, volumeToAttach.VolumeSpec), "volume_attach"),
```

Comment:
When CSI migration is enabled, the `volumeToAttach.VolumeSpec` will be the migrated CSI spec. However, with this change, it will be the original spec, since the migration logic is moved inside `attachVolumeFunc`. Im not sure if this is okay. Have you tested your change with CSI migration enabled? And can you check how is `CompleteFunc` is used?

Reply:
Do you mean a non-automated test? I'm quite new to here, How can I do that kind of test?

Reply:
We don't have automated test for CSI migration yet. You will need to turn on `CSIMigration` feature flag along with driver specific flag eg. (CSIMigrationAWS) when creating the cluster. Then deploy the CSI driver and verify that it still works.